base/compiler/ssair/passes.jl

```diff
@@ -953,14 +951,21 @@ function sroa_pass!(ir::IRCode, inlining::Union{Nothing,InliningState}=nothing)
             continue
         end
 
-        # analyze this `getfield` / `isdefined` / `setfield!` call
+        if is_finalizer
+            val = stmt.args[3]
+        else
+            # Analyze `getfield` / `isdefined` / `setfield!` call:
+
+            # Verify this statement won't throw, otherwise it can't be eliminated safely.
+            # XXX We should verify nothrow-ness of `getfield` statement here, but currently
+            # we are unable to prove it in the presence of potentially undefined fields.
+            # So we skip this verification for now since it would be proven to be nothrow
+            # if SROA on it is successful.
+            if !is_getfield && (compact[SSAValue(idx)][:flag] & IR_FLAG_NOTHROW) == 0
```

Comment:
I think we should probably still do the processing here so we can lift fields out of the sroa, and just not do the stmt removal at the end.

Reply:
In particular, it's quite possible that LLVM will be able to prove nothrow even if we can't, but it might not be able to do the SROA itself.

Reply:
Okay, I moved the check into `sroa_mutables!`.